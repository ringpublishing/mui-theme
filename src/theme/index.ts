export * from './theme';
export * from './config/breakpoints';
export * from './config/palette';
export * from './config/shape';
export * from './config/spacing';
export * from './config/typography';
