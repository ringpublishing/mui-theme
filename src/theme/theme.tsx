import {
    CssBaseline,
    PaletteMode,
    Theme,
    ThemeProvider,
    createTheme,
    PaletteColor,
    PaletteOptions
} from '@mui/material';
import { palette } from './config/palette';
import { breakpoints } from './config/breakpoints';
import { shape } from './config/shape';
import { typography } from './config/typography';
import { spacing } from './config/spacing';
import React from 'react';
import { CommonLanguages } from '../helpers/commonTypes';
import { plPL as corePL, enUS as coreUS } from '@mui/material/locale';

declare module '@mui/material/styles' {
    interface Palette {
        components: PaletteComponents;
        contrast: PaletteColor;
        common: CommonColors;
    }

    interface CommonColors {
        grey: string;
    }

    interface PaletteOptions {
        components: PaletteComponents;
        common?: Partial<CommonColors>;
    }

    interface PaletteComponents {
        datagrid: PaletteComponentDataGrid;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        [key: string]: any;
    }

    type PaletteComponentDataGrid = {
        border: string;
    };

    interface TypographyVariants {
        label: React.CSSProperties;
        headline1: React.CSSProperties;
        headline2: React.CSSProperties;
        headline3: React.CSSProperties;
    }

    interface TypographyVariantsOptions {
        label?: React.CSSProperties;
        headline1?: React.CSSProperties;
        headline2?: React.CSSProperties;
        headline3?: React.CSSProperties;
    }

    interface Theme {
        locale?: CommonLanguages;
    }
    interface ThemeOptions {
        locale?: CommonLanguages;
    }
}

declare module '@mui/material/Typography' {
    interface TypographyPropsVariantOverrides {
        label: true;
        headline1: true;
        headline2: true;
        headline3: true;
    }
}

declare module '@mui/material/Paper' {
    interface PaperPropsVariantOverrides {
        borderless: true;
    }
}

declare module '@mui/material/Button' {
    interface ButtonPropsColorOverrides {
        contrast: true;
    }
}

declare module '@mui/material/IconButton' {
    interface IconButtonPropsColorOverrides {
        contrast: true;
    }
}

export const getTheme = (
    mode: PaletteMode | string,
    language: CommonLanguages = CommonLanguages.enUS,
    externalComponentsTheme = {},
    externalLocales: object[] = []
): Theme => {
    const coreLocale = (language === CommonLanguages.plPL) ? corePL : coreUS;

    return createTheme({
        locale: language,
        palette: {
            ...palette[mode as PaletteMode] as PaletteOptions
        },
        breakpoints: {
            keys: ['xs', 'sm', 'md', 'lg', 'xl'],
            values: breakpoints
        },
        spacing,
        shape,
        typography,
        components: {
            ...externalComponentsTheme,
            MuiAutocomplete: {
                styleOverrides: {
                    root: ({ ownerState, theme }) => {
                        return {
                            ...(ownerState.multiple && {
                                '.MuiAutocomplete-endAdornment': {
                                    'height': `calc(100% - ${theme.spacing(1)})`,
                                    'display': 'flex',
                                    'align-items': 'end'
                                }
                            }),
                            '.MuiAutocomplete-clearIndicator': {
                                visibility: 'visible'
                            },
                            '.MuiInputBase-root': {
                                borderColor: 'red!important'
                            }
                        };
                    }
                }
            },
            MuiTooltip: {
                styleOverrides: {
                    tooltip: {
                        fontSize: '10px',
                        fontStyle: 'normal',
                        fontWeight: '400',
                        lineHeight: '14px'
                    }
                }
            },
            MuiPaper: {
                variants: [
                    {
                        props: { variant: 'borderless' },
                        style: {
                            '&.MuiPaper-borderless::before': {
                                height: 0
                            }
                        }
                    }
                ]
            },
            // Changes to the Accordion component to remove padding and margin and lower minHeight, so that the component can be used in a more compact way
            MuiAccordion: {
                styleOverrides: {
                    root: ({ theme }) => ({
                        'backgroundColor': 'transparent',
                        'minHeight': theme.spacing(4),
                        'margin': 0,
                        '&.Mui-expanded': {
                            margin: 0,
                            minHeight: theme.spacing(4)
                        }
                    })
                }
            },
            MuiAccordionSummary: {
                styleOverrides: {
                    root: ({ theme }) => ({
                        'padding': 0,
                        'minHeight': theme.spacing(4),
                        '&.Mui-expanded': {
                            padding: 0,
                            minHeight: theme.spacing(4)
                        }
                    }),
                    content: ({ theme }) => ({
                        'margin': theme.spacing(1, 0),
                        '&.Mui-expanded': {
                            margin: theme.spacing(1, 0)
                        }
                    })
                }
            },
            MuiAccordionDetails: {
                styleOverrides: {
                    root: ({ theme }) => ({
                        'padding': theme.spacing(1, 0),
                        '&.Mui-expanded': {
                            padding: theme.spacing(1, 0)
                        }
                    })
                }
            },
            /**
                 * Set the default variant for all possible components 'standard'.
                 * Components `MuiInputAdornment` and `MuiInputLabel` are not changed because they break components using variants other than 'standard'
                 * they inherit the variant from the parent component.
                 */
            MuiAlert: {
                defaultProps: {
                    variant: 'standard'
                }
            },
            MuiTextField: {
                defaultProps: {
                    variant: 'standard'
                }
            },
            MuiNativeSelect: {
                defaultProps: {
                    variant: 'standard'
                }
            },
            MuiFormControl: {
                defaultProps: {
                    variant: 'standard'
                }
            },
            MuiSelect: {
                defaultProps: {
                    variant: 'standard'
                }
            },
            MuiBadge: {
                defaultProps: {
                    variant: 'standard'
                }
            },
            MuiSvgIcon: {
                styleOverrides: {
                    fontSizeLarge: {
                        fontSize: '3.25rem'
                    }
                }
            },
            MuiAppBar: {
                styleOverrides: {
                    colorPrimary: ({ theme }) => {
                        return ({
                            backgroundColor: theme.palette.background.default
                        });
                    }
                }
            }
        }

    },
        coreLocale,
        ...externalLocales
    );
};

interface ThemeConfigProps {
    mode: PaletteMode;
    children: React.ReactNode[] | React.ReactNode;
    language?: CommonLanguages;
    externalLocales?: object[];
    externalComponentsTheme?: object;
}

export const ThemeConfig = ({ mode, children, language, externalLocales, externalComponentsTheme }: ThemeConfigProps): React.ReactNode[] | React.ReactNode => {
    return (
        <ThemeProvider theme={getTheme(mode, language, externalComponentsTheme, externalLocales)}>
            <CssBaseline />
            {children}
        </ThemeProvider>
    );
};

export const basicGrey100 = '#D9D9D9';
export const basicGrey200 = '#7b7b7b';
