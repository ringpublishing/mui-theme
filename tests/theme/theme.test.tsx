import { expect, it, describe, vi } from 'vitest';
import { getTheme, ThemeConfig } from '../../src/theme/theme';
import { render } from '@testing-library/react';
import { zhCN } from '@mui/material/locale';
import { plPL as xDataGridPl } from '@mui/x-data-grid/locales';
import { DataGrid } from '@mui/x-data-grid';
import { TablePagination } from '@mui/material';
import React from 'react';

describe('getTheme', () => {
    it('Minimum working example should work', () => {
        const themeObject = getTheme('light', 'plPL');
        expect(themeObject.locale).toBe('plPL');
        expect(themeObject.palette.mode).toBe('light');
        expect(themeObject.components).not.toHaveProperty('MuiDataGrid');
    });
    it('Full working example should work', () => {
        const themeObject = getTheme('dark', 'enUS', { MuiDataGrid: {
            defaultProps: {
                hideFooter: true
            },
            styleOverrides: {} } }, [zhCN, xDataGridPl]);
        expect(themeObject.locale).toBe('enUS');
        expect(themeObject.palette.mode).toBe('dark');
        expect(themeObject.components).toHaveProperty('MuiDataGrid');
    });
});

describe('ThemeConfig', () => {
    it('Minimum working example should work', () => {
        const { container } = render(<ThemeConfig mode={'light'} language={'enUS'}>
            <div>test</div>
        </ThemeConfig>);

        expect(container).toMatchSnapshot();
    });

    it('Example with chineese core locales and pl translations in datagrid', () => {
        const MuiDataGridPart = {
            MuiDataGrid: {
                defaultProps: {
                    hideFooter: true
                },
                styleOverrides: {
                    root: ({ theme }: {theme: Theme;}): CSS.Properties => {
                        return {
                            '& .MuiDataGrid-columnHeaderTitle': {
                                color: theme.palette.text.secondary
                            },
                            '--unstable_DataGrid-headWeight': 600,
                            '--DataGrid-rowBorderColor': theme.palette.components.datagrid.border,
                            'border': 'none',
                            '& .MuiDataGrid-cell:focus': {
                                outline: 'none'
                            },
                            '& .MuiDataGrid-cell:focus-within': {
                                outline: 'none'
                            },
                            '& .MuiDataGrid-columnHeader:focus': {
                                outline: 'none'
                            },
                            '& .MuiDataGrid-columnHeader:focus-within': {
                                outline: 'none'
                            },
                            '& .MuiDataGrid-row:hover': {
                                cursor: 'pointer'
                            }
                        };
                    }
                }
            }
        };
        const { container } = render(<ThemeConfig mode={'light'} language={'enUS'} externalComponentsTheme={MuiDataGridPart}
            externalLocales={[xDataGridPl, zhCN]}
        >

            <div style={{ height: '400px' }} >
                <DataGrid
                    rows={[]}
                    columns={[{
                        field: 'firstName',
                        headerName: 'First name',
                        width: 150,
                        editable: true
                    }]}
                />
                <TablePagination
                    count={2000}
                    rowsPerPage={10}
                    page={1}
                    component="div"
                    onPageChange={vi.fn()}
                />
            </div>
        </ThemeConfig>);

        expect(container).toMatchSnapshot();
    });
});
