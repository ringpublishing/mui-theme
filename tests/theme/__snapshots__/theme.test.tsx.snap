// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeConfig > Example with chineese core locales and pl translations in datagrid 1`] = `
<div>
  <div
    style="height: 400px;"
  >
    <div
      class="MuiDataGrid-root MuiDataGrid-root--densityStandard MuiDataGrid-root--noToolbar MuiDataGrid-withBorderColor css-vr43xx-MuiDataGrid-root"
      style="--DataGrid-width: 0px; --DataGrid-hasScrollX: 1; --DataGrid-hasScrollY: 0; --DataGrid-scrollbarSize: 0px; --DataGrid-rowWidth: 150px; --DataGrid-columnsTotalWidth: 150px; --DataGrid-leftPinnedWidth: 0px; --DataGrid-rightPinnedWidth: 0px; --DataGrid-headerHeight: 56px; --DataGrid-headersTotalHeight: 56px; --DataGrid-topContainerHeight: 56px; --DataGrid-bottomContainerHeight: 0px; --height: 52px;"
    >
      <div
        aria-colcount="1"
        aria-multiselectable="false"
        aria-rowcount="1"
        class="MuiDataGrid-main css-3eek4p-MuiDataGrid-main"
        role="grid"
        tabindex="-1"
      >
        <div
          class="css-qtdhae"
          data-id="gridPanelAnchor"
          role="presentation"
        />
        <div
          class="MuiDataGrid-virtualScroller css-1793420-MuiDataGrid-virtualScroller"
          role="presentation"
          style="overflow-x: hidden;"
        >
          <div
            class="MuiDataGrid-topContainer MuiDataGrid-container--top css-2sp4oa"
            role="presentation"
          >
            <div
              class="MuiDataGrid-columnHeaders css-1pmk00y-MuiDataGrid-columnHeaders"
              role="presentation"
            >
              <div
                aria-rowindex="1"
                class="MuiDataGrid-row--borderBottom css-yrdy0g-MuiDataGrid-columnHeaderRow"
                role="row"
              >
                <div
                  role="presentation"
                  style="width: 0px;"
                />
                <div
                  aria-colindex="1"
                  aria-label="First name"
                  aria-sort="none"
                  class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable MuiDataGrid-withBorderColor MuiDataGrid-columnHeader--last"
                  data-field="firstName"
                  role="columnheader"
                  style="height: 56px; width: 150px; min-width: 150px; max-width: 150px;"
                  tabindex="0"
                >
                  <div
                    class="MuiDataGrid-columnHeaderDraggableContainer"
                    draggable="false"
                    role="presentation"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainer"
                      role="presentation"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainerContent"
                      >
                        <div
                          aria-label=""
                          class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                          data-mui-internal-clone-element="true"
                        >
                          First name
                        </div>
                      </div>
                      <div
                        class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                      >
                        <button
                          aria-label="Sortuj"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1dw0joi-MuiButtonBase-root-MuiIconButton-root"
                          field="firstName"
                          tabindex="-1"
                          title="Sortuj"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-1cea49u-MuiSvgIcon-root"
                            data-testid="ArrowUpwardIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-menuIcon"
                    >
                      <button
                        aria-expanded="false"
                        aria-haspopup="menu"
                        aria-label="Menu"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1dw0joi-MuiButtonBase-root-MuiIconButton-root"
                        data-mui-internal-clone-element="true"
                        id=":r4:"
                        tabindex="-1"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-1cea49u-MuiSvgIcon-root"
                          data-testid="TripleDotsVerticalIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                          />
                        </svg>
                        <span
                          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--resizable MuiDataGrid-columnSeparator--sideRight"
                    style="min-height: 56px; opacity: 1;"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-1rrw7xm-MuiSvgIcon-root"
                      data-testid="SeparatorIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11 19V5h2v14z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="MuiDataGrid-filler"
                  role="presentation"
                />
                <div
                  class="MuiDataGrid-scrollbarFiller MuiDataGrid-scrollbarFiller--header"
                  role="presentation"
                />
              </div>
            </div>
          </div>
          <div
            class="MuiDataGrid-overlayWrapper css-74bi4q-MuiDataGrid-overlayWrapper"
          >
            <div
              class="MuiDataGrid-overlayWrapperInner css-1akuw9y-MuiDataGrid-overlayWrapperInner"
              style="height: -56px; width: 0px;"
            >
              <div
                class="MuiDataGrid-overlay css-1w53k9d-MuiDataGrid-overlay"
              >
                Brak danych
              </div>
            </div>
          </div>
          <div
            class="MuiDataGrid-virtualScrollerContent css-1kwdphh-MuiDataGrid-virtualScrollerContent"
            role="presentation"
            style="width: auto; height: 0px;"
          >
            <div
              class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
              role="rowgroup"
              style="transform: translate3d(0, 0px, 0);"
            />
          </div>
          <div
            class="MuiDataGrid-bottomContainer MuiDataGrid-container--bottom css-14egwce"
            role="presentation"
          />
        </div>
        <div
          aria-hidden="true"
          class="MuiDataGrid-scrollbar MuiDataGrid-scrollbar--horizontal css-1rtad1"
          tabindex="-1"
        >
          <div
            class="MuiDataGrid-scrollbarContent css-1baulvz"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiTablePagination-root css-19eh4pg-MuiTablePagination-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-mho382-MuiToolbar-root-MuiTablePagination-toolbar"
      >
        <div
          class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
        />
        <p
          class="MuiTablePagination-selectLabel css-vg9wej-MuiTablePagination-selectLabel"
          id=":r1:"
        >
          每页行数:
        </p>
        <div
          class="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-input css-1ned3pt-MuiInputBase-root-MuiTablePagination-select"
        >
          <div
            aria-controls=":r2:"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby=":r1: :r0:"
            class="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-d5l9zj-MuiSelect-select-MuiInputBase-input"
            id=":r0:"
            role="combobox"
            tabindex="0"
          >
            10
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="10"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-19deldj-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <p
          class="MuiTablePagination-displayedRows css-qbny2j-MuiTablePagination-displayedRows"
        >
          第 11 条到第 20 条，共 2000 条
        </p>
        <div
          class="MuiTablePagination-actions"
        >
          <button
            aria-label="上一页"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-9q913r-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            title="上一页"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1rrw7xm-MuiSvgIcon-root"
              data-testid="KeyboardArrowLeftIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="下一页"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-9q913r-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            title="下一页"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1rrw7xm-MuiSvgIcon-root"
              data-testid="KeyboardArrowRightIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ThemeConfig > Minimum working example should work 1`] = `
<div>
  <div>
    test
  </div>
</div>
`;
